helpers/spark-shell.py

```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env python3
+
+import os
+import pwd
+import logging
+import argparse
+import utils
+
+USER_HOME_DIR_ENT_IDX = 5
+
+if __name__ == "__main__":
+    logging.basicConfig(format='%(asctime)s %(levelname)s %(message)s', level=logging.INFO)
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--master", default=None, type=str, help='Kubernetes control plane uri.')
+    parser.add_argument("--deploy-mode", default="client", type=str, help='Deployment mode for spark shell. Default is \'client\'.')
+    parser.add_argument("--properties-file", default=None, type=str, help='Spark default configuration properties file.')
+    args, extra_args = parser.parse_known_args()
+
+    os.environ["HOME"] = pwd.getpwuid(os.getuid())[USER_HOME_DIR_ENT_IDX]
+    if os.environ.get('SPARK_HOME') is None or os.environ.get('SPARK_HOME') == '':
+        os.environ['SPARK_HOME'] = os.environ['SNAP']
+
+    STATIC_DEFAULTS_CONF_FILE = utils.get_static_defaults_conf_file()
+    DYNAMIC_DEFAULTS_CONF_FILE = utils.get_dynamic_defaults_conf_file()
+    ENV_DEFAULTS_CONF_FILE = utils.get_env_defaults_conf_file()
+
+    snap_static_defaults = utils.read_property_file(STATIC_DEFAULTS_CONF_FILE)
+    setup_dynamic_defaults = utils.read_property_file(DYNAMIC_DEFAULTS_CONF_FILE) if os.path.isfile(DYNAMIC_DEFAULTS_CONF_FILE) else dict()
+    env_defaults = utils.read_property_file(ENV_DEFAULTS_CONF_FILE) if ENV_DEFAULTS_CONF_FILE and os.path.isfile(ENV_DEFAULTS_CONF_FILE) else dict()
+    props_file_arg_defaults = utils.read_property_file(args.properties_file) if args.properties_file else dict()
+
+    with utils.UmaskNamedTemporaryFile(mode = 'w', prefix='spark-conf-', suffix='.conf') as t:
+        defaults = utils.merge_configurations([snap_static_defaults, setup_dynamic_defaults, env_defaults, props_file_arg_defaults])
+        SCALA_HISTORY_FILE = utils.get_scala_shell_history_file()
+        defaults['spark.driver.extraJavaOptions'] = f'-Dscala.shell.histfile={SCALA_HISTORY_FILE}'
```

Comment:
what does it happen if this key is already present? Can we put some logic to merge the two configurations for the extraJavaOptions rather than overwriting them?

I suppose we could see the use case where the user specifies some parameters that then gets overridden by this line. 

BTW...it would be nice to carry that merging logic also to the other parts of the merge_configurations. But let's discuss this a bit more before addressing this. Here I'm just mentioning to start to think about this

Reply:
Yeah, this does not look right. 

Instead, I guess I can do this immediately after getting setup related dynamic defaults.

dynamic_defaults['spark.driver.extraJavaOptions'] = f'-Dscala.shell.histfile={SCALA_HISTORY_FILE}'

This way it can be overridden with env file and properties file that are parsed next in order.